text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,100 @@
+2. Omitted `appliesTo` means that the action cannot be used in a request. It is used exclusively as an action group.
+3. Disallow empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. Disallow an empty `appliesTo` attribute thus requiring at least one of `principalTypes`, `resourceTypes` or `context` to be specified if `appliesTo` is specified.
+
+This feature request is consistent with the semantics for `appliesTo` proposed in [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) for the alternate Cedar schema syntax.
+
+## Detailed design
+
+This RFC will require the following changes:
+
+1. The permissive validator must be updated to treat an omitted `appliesTo.principalTypes` / `appliesTo.resourceTypes` as the _unspecified_ entity type, rather than _any_ entity type. The difference is that the _unspecified_ entity type is a particular type that is different from any defined type, whereas _any_ entity type could be the unspecified type or any other type defined in the schema. Note that this change is only required in the _permissive_ validator, because the corresponding change was already made to strict validation as part of [RFC 19](https://github.com/cedar-policy/rfcs/blob/main/text/0019-stricter-validation.md).
+2. The validator must be updated to treat an omitted `appliesTo` attribute as indication that an action does not apply to any principal or resource types.
+3. The schema parser must be updated to reject empty arrays for `appliesTo.principalTypes` / `appliesTo.resourceTypes`.
+4. The schema parser must be updated to reject empty `appliesTo` attributes.
+
+### Notes on Backwards Compatibility
+
+This RFC proposes a **breaking change** that requires a major version bump to Cedar.
+For the sake of argument, say that the current version of Cedar X and that this RFC will be implemented in Cedar Y (which has no other breaking changes).
+The rest of this section is written for users who are upgrading from X to Y.
+
+The changes proposed in this RFC will _only_ affect you if you have one or more of the following in your schema:
+
+1. You intend for an action to apply to _no_ principals or resources, i.e.,
+
+- `appliesTo.principalTypes: []`
+- `appliesTo.resourceTypes: []`
+
+2. You intend for an action to apply to _any_ principal and/or resource types, i.e.,
+
+- omitted `appliesTo` field
+- omitted `appliesTo.principalTypes` field
+- omitted `appliesTo.resourceTypes` field
+
+In **case 1**, in preparation to upgrade, you will need to update your schema to omit the `appliesTo` field.
+Your current schema will result in error when you upgrade to Y.
+If you use your modified schema with X, you will be able to validate policies that will not validate in Y.
+There is _no way_ to write a schema so that validation behavior will be the same between X and Y.
```

Comment:
starting a different thread based on the same RFC text about backwards compatibility:

One way to mitigate some of these backwards compatibility problems would be to combine this with [RFC 33](https://github.com/cedar-policy/rfcs/pull/33).  For instance, at least two options:
* force Cedar Y schemas to include an explicit version marker saying version Y.  Schemas with no version marker will be interpreted using the Cedar X behavior, even if you're using Cedar Y.  This gives full backwards-compatibility: no one's schema changes behavior when upgrading from Cedar X to Cedar Y, unless/until you add a version marker saying version Y.  However, this means the behavior proposed in this RFC won't be default, which seems undesirable.
* same thing, but interpret schemas with no version marker using the Cedar Y behavior.  This is still a breaking change, because unmarked schemas will change behavior when upgrading from Cedar X to Cedar Y.  But it solves the problem of "no way to write a schema so that validation behavior will be the same between X and Y."  Users just add a "version X" marker to their schema, and then they can upgrade from Cedar X to Cedar Y without that schema changing behavior.

Of these two options, I prefer the latter because it makes this RFC's proposed behavior the default.  But both of these options of course come with the caveats/downsides discussed on RFC 33.

Reply:
If someone out there was relying on schemas, and they were storing a creationDate and lastModifiedDate, they could use that information to decide which version to use in schema parsing and validation. If the lastModifiedDate < some cutoff, then use Cedar X. Otherwise use Cedar Y. I think craig is on to something here.

Reply:
Keep in mind that for RFC33 the original proposal was that later versions of the software will not (necessarily) accept earlier versions of the inputs. You are suggesting to that later software versions interpret earlier input versions ("Schemas with no version marker will be interpreted using the Cedar X behavior, even if you're using Cedar Y."). 

We didn't want to commit to doing this in general. If we are not going to do it, I realize that versions could just be stored as metadata in whatever way an application likes, rather than in policies/JSON/etc so that Cedar's tools have to specifically know about it. In particular, we could just add a general file-wide annotation mechanism that generalizes the per-policy mechanism we have now, and apps can use it to add version info however they like.

In sum: We should only add versioning info to Cedar itself if we are going to have later versions of tools/code interpret out-of-date versions of inputs (which Craig is proposing to do here). But agreeing to understand past input versions complicates development, testing, etc. so we should be really sure about what we are getting into.

Reply:
If we introduce another breaking change to the schema in Cedar Z, then how do we handle the schemas in versions X,Y, and Z? Can Y schema just delete the version Y so that it will be parsed with the Cedar Y? What will happen to the X schemas with no version marker?

Reply:
> If we introduce another breaking change to the schema in Cedar Z

then Cedar Z would have to handles schemas from X, Y, and Z.  This is why @mwhicks1 calls this a slippery slope and says "we [don't] want to commit to doing this in general".

> Can Y schema just delete the version Y so that it will be parsed with Cedar Y?

I assume you're referring to the second bullet of my proposal.  If so, then yes, but it wouldn't be required.

> What will happen to the X schemas with no version marker?

in the first bullet of my proposal, they would not break.  in the second bullet of my proposal, they would.

Reply:
(of course, they only break if they meet the conditions described in the RFC's backwards compatibility section.  Many schemas are not affected by RFC 35)

Reply:
I'm 💯 on board with @cdisselkoen's second bullet. I also like the first bullet, but based on discussion on [RFC33](https://github.com/cedar-policy/rfcs/pull/33) I've been convinced that leaving version info optional is cleanest, which suggests that a schema without a version should be interpreted as the latest version, rather than always always defaulting to an old version (e.g., 2.4).

To @mwhicks1's point on the slipper slope: In most cases, I think we'd only want to use the version info from RFC33 to print better error messages (e.g., "version X is no longer supported, please follow the upgrade instructions at < some link >"). I think we should treat the current RFC (35), where we've changed the interpretation of the same syntax between versions, as a special case and provide support for parsing the old format for a period of time (e.g., until the next major release).

Reply:
> I think we should treat the current RFC (35), where we've changed the interpretation of the same syntax between versions, as a special case and provide support for parsing the old format for a period of time (e.g., until the next major release).

Surely this will not be the last time that a situation like this arises. So we are setting a precedent here.

Reply:
> Surely this will not be the last time that a situation like this arises. So we are setting a precedent here.

I share @mwhicks1's concern:  we're setting a precedent here.

Reply:
I agree that this is not the last time this will happen, and we're setting a precedent.

What I meant was that I think it's ok to provide support for old versions in cases where we have made a breaking change like this -- but that we should set the precedent that the old version is not supported indefinitely. E.g., we would only support it until the next major version.

Reply:
Updated the RFC text to include this alternative, and started a fresh thread for discussion: https://github.com/cedar-policy/rfcs/pull/35#discussion_r1386794836

Resolving this thread to keep the discussion organized.